### ISSUE TITLE
Fail to load transform when `scale_to_z_score_per_key` is used

### ISSUE THREAD

**Issue by rafaelbarreto87** [tensorflow/transform#174]
Hey team,

I have a couple of weighted sparse features in the form of:

```
{
    'feat_keys': tf.io.VarLenFeature(tf.string),
    'feat_weights': tf.io.VarLenFeature(tf.float32),
}
```

`feat_keys` and `feat_weights` are exactly the same dimension. In my case, the `feat_weights` is the value of multiple decayed counts and the `feat_keys` indicates the decay. Since each `feat_weights` entry correspond to a count distinct from the other values, I'm trying to use `scale_to_z_score_per_key` to normalize them. My pre-processing function looks like this:

```
def preprocessing_fn(feats):
    feats = dict(feats)

    for f in FEATS_TO_BUILD_VOCAB:
        tft.vocabulary(feats[f], vocab_filename=f)

    for k, v in CONTS_FEATS_TO_NORMALIZE_BY_KEY.items():
        feats[v + '__norm'] = tft.scale_to_z_score_per_key(
            x=feats[v],
            key=feats[k],
        )

    return feats
```

The transform is produced as expected. However, when I try to use it, I get this error:

```
InvalidArgumentError: Op type not registered
'scale_to_z_score_per_key_mean_and_var_per_key_assert_equal_1_Assert_AssertGuard_true_251'
in binary running on notebook-de-gpu-0. Make sure the Op and Kernel are registered in the
binary running in this process. Note that if you are loading a saved graph which used ops from
tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as
contrib ops are lazily registered when the module is first accessed. while building NodeDef
'model_1/transform_features_layer_1/transform/transform/scale_to_z_score_per_key/mean_and_var_per_key/assert_equal_1/Assert/AssertGuard/then/_0'
[Op:__inference_distributed_function_5652]
```

It looks that some parts of the transform are not available in the binary somehow.

I thought I was making some basic mistake in how I'm using TF Transform, but if I comment out the `tft.scale_to_z_score_per_key` piece, all works as expected, so this looks like a bug. Any ideas of what might be happening and how I can debug this?

I'm using TF 2.1 (eager mode) and TFT 0.21.2.

**Comment by rafaelbarreto87** [comment]
After debugging this a bit more, it seems assertions in the preprocessing function are the problem somehow. The following dummy preprocessing function exhibits the same behavior:

```
def preprocessing_fn(feats):
    feats = dict(feats)

    for k, v in CONTS_FEATS_TO_NORMALIZE_BY_KEY.items():
        x = feats[k].dense_shape
        y = feats[v].dense_shape
        with tf.control_dependencies([tf.compat.v1.assert_equal(x, y)]):            
            pass

    return feats
```

Analyze phase succeeds without any problem, but when I call `transform_raw_features` in a dataset, it fails. Specifically, if I run the following

```
tft_transform_output = tft.TFTransformOutput(TRANSFORM_FN_DIR)

def get_dataset(file_pattern, batch_size, shuffle=True, n_epochs=None):

    def process_features(feats):
        return tft_transform_output.transform_raw_features(feats)

    return tf.data.experimental.make_batched_features_dataset(
        file_pattern=file_pattern,
        batch_size=batch_size,
        features=tft_transform_output.raw_feature_spec(),
        shuffle=shuffle,
        num_epochs=n_epochs,
        reader_args=['GZIP'],
        sloppy_ordering=True
    ).map(process_features)

for b in get_dataset(training_files, 256, shuffle=True, n_epochs=None).take(1):
    pass
```

I get the error: ```InvalidArgumentError: Unable to find FunctionDef for assert_equal_1_Assert_AssertGuard_false_248 in the registry. [Op:OptimizeDataset]```.

It might be that there's an issue in the way the graph is exported somehow?

**Comment by rafaelbarreto87** [comment]
It looks like this is a more general problem indeed. I'm encountering this issue now with the following code:

```
def preprocessing_fn(feats):
    feats = dict(feats)
    ...
    feats['unix_ts_local'] = tf.sparse.add(
        feats['unix_ts_utc'],
        tf.sparse.reset_shape(feats['tz_offset_in_secs'], feats['unix_ts_utc'].dense_shape)
    )
    ...
    return feats
```

The cultprit is `tf.sparse.reset_shape`. When I use it, the Analyze step succeeds without any problems. But when I try to use it in a dataset, I get:

```
InvalidArgumentError: Unable to find FunctionDef for assert_less_equal_Assert_AssertGuard_true_233117 in the registry. [Op:OptimizeDataset]
```

I'm using Tensorflow 2.1. I don't know if this has anything to do with it.

**Comment by rafaelbarreto87** [comment]
Ops. I don't know what I did, but @gowthamkpr was unassigned. Will assign it again.

**Comment by zoyahav** [comment]
@rafaelbarreto87 how are you using Transform? I'm curious about the eager part specifically as Transform doesn't run in an eager context.
This could be an issue of TF control dependencies v1 vs. v2.

**Comment by rafaelbarreto87** [comment]
@zoyahav What I meant is that I'm using TF 2.1 and, hence, eager mode is the default. Here's an example of how I'm using it:

```
tft_transform_output = tft.TFTransformOutput(TRANSFORM_FN_DIR)


def get_dataset(file_pattern, batch_size, shuffle=True, n_epochs=None):

    def process_features(feats):
        return tft_transform_output.transform_raw_features(feats)

    return tf.data.experimental.make_batched_features_dataset(
        file_pattern=file_pattern,
        batch_size=batch_size,
        features=tft_transform_output.raw_feature_spec(),
        shuffle=shuffle,
        num_epochs=n_epochs,
        reader_args=['GZIP'],
        sloppy_ordering=True
    ).map(process_features)


for b in get_dataset(training_files, 256, shuffle=True, n_epochs=None).take(1):
    pass  # it crashes in the first iteration
```

**Comment by rafaelbarreto87** [comment]
I just tried this in Tensorflow 1.15 with eager mode enabled as well (i.e. `tf.compat.v1.enable_eager_execution()`) and it doesn't work either. However, I get a different exception:

```
InternalError: {{function_node __inference_Dataset_map_TFTransformOutput.transform_raw_features_48586}} Retval[0] does not have value [Op:IteratorGetNextSync]
```

Maybe the issue is eager mode?

**Comment by zoyahav** [comment]
Could you post you entire pipeline?
There are still some missing pieces here that we would need in order to diagnose.

**Comment by rafaelbarreto87** [comment]
I tried to get to a bare bones example here. I had to remove some parameters and dataset names since this is internal code, but here it is:

```
import datetime
from functools import reduce
import json
import logging
import os
os.environ['CUDA_VISIBLE_DEVICES']= '-1'
import random

import apache_beam as beam
from google.protobuf import text_format
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.python.lib.io import file_io
import tensorflow_transform as tft
import tensorflow_transform.beam as tft_beam
from tensorflow_transform.tf_metadata import dataset_metadata
from tensorflow_transform.tf_metadata import schema_utils


logging.getLogger().setLevel(logging.ERROR)
tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)


PROJECT = !gcloud config get-value project
PROJECT = PROJECT[0]
BASE_GCS_PATH = 'gs://{}/workspaces/rbarreto'.format(PROJECT)
DATASET_BASE_PATH = os.path.join(BASE_GCS_PATH, 'datasets', 'my_dataset')
TRANSFORM_FN_DIR = os.path.join(DATASET_BASE_PATH, 'tft', 'tmp')
OVERWRITE_TRANSFORM = True


def get_tfrecords_glob(ts, dt_start, dt_end):
    include_all_days = dt_start.month < ts.month < dt_end.month
    include_all_hours = dt_start <= ts.floor('1d') < dt_end
    return os.path.join(
        os.path.join(DATASET_BASE_PATH, 'data'),
        '{:04d}'.format(ts.year),
        '{:02d}'.format(ts.month),
        '*' if include_all_days else '{:02d}'.format(ts.day),
        '*' if include_all_hours  else '{:02d}'.format(ts.hour),
        '*.tfrecord.gz'
    )


dt_start = pd.Timestamp('5/18/2020')
dt_end = pd.Timestamp('5/19/2020')
training_files = sorted(list({
    get_tfrecords_glob(ts, dt_start, dt_end)
    for ts in pd.date_range(start=dt_start, end=dt_end, freq='H', closed='left')
}))
raw_examples_schema = text_format.Parse(
    message=tft.tf_metadata.schema_utils.schema_pb2.Schema(),
    text=file_io.read_file_to_string(os.path.join(DATASET_BASE_PATH, 'schema.pbtxt'))
)
raw_examples_coder = tft.coders.ExampleProtoCoder(raw_examples_schema)
raw_examples_metadata = dataset_metadata.DatasetMetadata(raw_examples_schema)


CONTS_FEATS_TO_NORMALIZE_BY_KEY = {
    'app_engagement_counts__keys': 'app_engagement_counts__values'
}


def preprocessing_fn(feats):
    feats = dict(feats)
    
    for k, v in CONTS_FEATS_TO_NORMALIZE_BY_KEY.items():
        x = feats[k].dense_shape
        y = feats[v].dense_shape
        with tf.control_dependencies([tf.compat.v1.assert_equal(x, y)]):            
            pass
        
    return feats


pipeline_options_kwargs = {
    'job_name': 'my-tft-{}'.format(
        datetime.datetime.utcnow().strftime('%Y%m%d%H%M%S')
    ),
    'runner': 'DataflowRunner',
    'project': PROJECT,
    'region': 'us-central1',
    'temp_location': os.path.join(BASE_GCS_PATH, 'temp'),
    'save_main_session': True,
    'service_account_email': 'MY_SERVICE_ACCOUNT',
    'autoscaling_algorithm': 'NONE',
    'num_workers': 10,
    'machine_type': 'n1-standard-16',
    'setup_file': './dataflow_dependencies/setup.py',
}
pipeline_options = beam.pipeline.PipelineOptions(flags=[], **pipeline_options_kwargs)
with beam.Pipeline(options=pipeline_options) as pipeline:
    with tft_beam.Context(temp_dir=pipeline_options_kwargs['temp_location']):
        raw_examples_pcolls = []
        for i, pattern in enumerate(training_files):
            raw_examples_pcolls.append(
                pipeline | ('ReadRawExamples_' + str(i)) >> beam.io.ReadFromTFRecord(
                    pattern, coder=raw_examples_coder
                )
            )
        raw_examples = raw_examples_pcolls | beam.Flatten()
        raw_examples_dataset = (raw_examples, raw_examples_metadata)
        raw_examples_metadata | 'WriteRawMetadata' >> tft_beam.WriteMetadata(
            os.path.join(TRANSFORM_FN_DIR, tft.TFTransformOutput.RAW_METADATA_DIR),
            pipeline=pipeline
        )
        transform_fn = raw_examples_dataset | tft_beam.AnalyzeDataset(preprocessing_fn)
        transform_fn | 'WriteTransformFn' >> tft_beam.WriteTransformFn(TRANSFORM_FN_DIR)


tft_transform_output = tft.TFTransformOutput(TRANSFORM_FN_DIR)


def get_dataset(file_pattern, batch_size):
    return tf.data.experimental.make_batched_features_dataset(
        file_pattern=file_pattern,
        batch_size=batch_size,
        features=tft_transform_output.raw_feature_spec(),
        reader_args=['GZIP'],
    ).map(tft_transform_output.transform_raw_features)


for i, b in enumerate(get_dataset(training_files, 1).take(100)):  # ensure transform actually works
    # exception happens here
    pass
```

**Comment by rafaelbarreto87** [comment]
In case anyone arrived here with the same problem, I was able to implement the scaling logic using `tft.ptransform_analyzer`. Here's the gist: https://gist.github.com/rafaelbarreto87/1670426783b046ffff17c45dd063f45e. This was put together in hasty and I'm sure it's not the most optimal or clean way of doing it. However, it works.

**Comment by rafaelbarreto87** [comment]
This problem is now showing up when using RaggedTensors. I have the following operation as part of the transform

```
def does_intersect(set1, set2):
    return transform_sparse_values(
        tf.RaggedTensor.from_sparse(tf.sets.intersection(set1, set2))[:, :1].to_sparse(),
        lambda v: tf.ones_like(v, dtype=tf.int64),
    )
```

The transform is generated correctly, but it can't be loaded. It throws:

```
InvalidArgumentError: Unable to find FunctionDef for RaggedFromSparse_Assert_AssertGuard_true_601 in the registry. [Op:OptimizeDataset]
```

Updating to TF2.3.1 and TF Transform 0.24.1 does not solve it.

Unfortunately this is blocking at this point and I don't know how to proceed. Is there any recommendation on what to do?

**Comment by zoyahav** [comment]
You shouldn't need to implement a ptransform_analyzer for this issue, please help us debug this issue by providing a minimal repro that we would be able to run and debug (one that doesn't need to read data that's not available).

For the initial issue with asserts, I've tried reproducing with a stand-alone pipeline in colab, but I'm not getting the issue:

```
import pprint
import tempfile


import tensorflow as tf
import tensorflow_transform as tft
import tensorflow_transform.beam.impl as tft_beam
from tensorflow_transform.tf_metadata import dataset_metadata
from tensorflow_transform.tf_metadata import schema_utils

print(tf.__version__)
print(tft.__version__)

def preprocessing_fn(feats):
    feats = dict(feats)
    
    x = feats['x'].dense_shape
    y = feats['y'].dense_shape
    with tf.control_dependencies([tf.compat.v1.assert_equal(x, y)]):            
        pass
        
    return feats

raw_data = [
    {'x': [1], 'y': [1], 's': ['hello']},
    {'x': [2], 'y': [2], 's': ['world']},
    {'x': [3], 'y': [3], 's': ['hello']}
]

raw_data_metadata = dataset_metadata.DatasetMetadata(
    schema_utils.schema_from_feature_spec({
        's': tf.VarLenFeature(tf.string),
        'y': tf.VarLenFeature(tf.float32),
        'x': tf.VarLenFeature(tf.float32),
    }))

with tft_beam.Context(temp_dir=tempfile.mkdtemp()):
  transformed_dataset, transform_fn = (  # pylint: disable=unused-variable
      (raw_data, raw_data_metadata) | tft_beam.AnalyzeAndTransformDataset(
          preprocessing_fn))

transformed_data, transformed_metadata = transformed_dataset  # pylint: disable=unused-variable

pprint.pprint(transformed_data)
```

Output:

> 1.15.2
> 0.24.1
> WARNING:tensorflow:You are passing instance dicts and DatasetMetadata to TFT which will not provide optimal performance. Consider following the TFT guide to upgrade to the TFXIO format (Apache Arrow RecordBatch).
> WARNING:tensorflow:You are passing instance dicts and DatasetMetadata to TFT which will not provide optimal performance. Consider following the TFT guide to upgrade to the TFXIO format (Apache Arrow RecordBatch).
> INFO:tensorflow:Assets added to graph.
> INFO:tensorflow:Assets added to graph.
> INFO:tensorflow:No assets to write.
> INFO:tensorflow:No assets to write.
> INFO:tensorflow:SavedModel written to: /tmp/tmp2t4xpa4o/tftransform_tmp/7189755b3d894895a03930104a277a4f/saved_model.pb
> INFO:tensorflow:SavedModel written to: /tmp/tmp2t4xpa4o/tftransform_tmp/7189755b3d894895a03930104a277a4f/saved_model.pb
> WARNING:tensorflow:You are passing instance dicts and DatasetMetadata to TFT which will not provide optimal performance. Consider following the TFT guide to upgrade to the TFXIO format (Apache Arrow RecordBatch).
> WARNING:tensorflow:You are passing instance dicts and DatasetMetadata to TFT which will not provide optimal performance. Consider following the TFT guide to upgrade to the TFXIO format (Apache Arrow RecordBatch).
> INFO:tensorflow:Saver not created because there are no variables in the graph to restore
> INFO:tensorflow:Saver not created because there are no variables in the graph to restore
> INFO:tensorflow:Saver not created because there are no variables in the graph to restore
> INFO:tensorflow:Saver not created because there are no variables in the graph to restore
> [{'s': array([b'hello'], dtype=object),
>   'x': array([1.], dtype=float32),
>   'y': array([1.], dtype=float32)},
>  {'s': array([b'world'], dtype=object),
>   'x': array([2.], dtype=float32),
>   'y': array([2.], dtype=float32)},
>  {'s': array([b'hello'], dtype=object),
>   'x': array([3.], dtype=float32),
>   'y': array([3.], dtype=float32)}]


For the second issue with RaggedTensor we'll also need a repro to proceed.

**Comment by rafaelbarreto87** [comment]
This is right now part of a bigger pipeline. I'll if I can create a self-contained example.

By the error, I believe this only happens when TF datasets are involved: `Unable to find FunctionDef for RaggedFromSparse_Assert_AssertGuard_false_592 in the registry. [Op:OptimizeDataset]`. Note the `[Op:OptimizeDataset]`.

I'll create some example using a public dataset and try to reproduce it there.

**Comment by rafaelbarreto87** [comment]
Ok. This should do it:

```
import pprint
import tempfile

import apache_beam as beam
import tensorflow as tf
import tensorflow_transform as tft
import tensorflow_transform.beam as tft_beam
from tensorflow_transform.tf_metadata import dataset_metadata
from tensorflow_transform.tf_metadata import schema_utils


print(tf.__version__)
print(tft.__version__)


transform_dir = tempfile.mkdtemp()


def preprocessing_fn(feats):
    feats = dict(feats)
    
    feats['broken_feat'] = tf.RaggedTensor.from_sparse(feats['feature2'])[:, :1].to_sparse()
        
    return feats

raw_data = [
    {'feature0': [1], 'feature1': [1], 'feature2': ['hello'], 'feature3': [2.0]},
    {'feature0': [1], 'feature1': [1], 'feature2': ['hello'], 'feature3': [3.0]}
]

feature_description = {
    'feature0': tf.io.VarLenFeature(tf.int64),
    'feature1': tf.io.VarLenFeature(tf.int64),
    'feature2': tf.io.VarLenFeature(tf.string),
    'feature3': tf.io.VarLenFeature(tf.float32),
}

raw_data_metadata = dataset_metadata.DatasetMetadata(
    schema_utils.schema_from_feature_spec(feature_description)
)

pipeline_options = beam.pipeline.PipelineOptions(flags=[], runner='DirectRunner')
with beam.Pipeline(options=pipeline_options) as pipeline:
    with tft_beam.Context(temp_dir=tempfile.mkdtemp()):
        raw_data_metadata | 'WriteRawMetadata' >> tft_beam.WriteMetadata(
            os.path.join(transform_dir, tft.TFTransformOutput.RAW_METADATA_DIR),
            pipeline=pipeline
        )
        transform_fn = (raw_data, raw_data_metadata) | tft_beam.AnalyzeDataset(preprocessing_fn)
        transform_fn | 'WriteTransformFn' >> tft_beam.WriteTransformFn(transform_dir)
        
        
## create a fake dataset with an encoded tf.Example with the same schema

tf_examples_ds = tf.data.Dataset.from_tensor_slices(
    tf.constant([b'\nQ\n\x13\n\x08feature2\x12\x07\n\x05\n\x03cat\n\x14\n\x08feature3\x12\x08\x12\x06\n\x04J\xec\x9a\xbd\n\x11\n\x08feature0\x12\x05\x1a\x03\n\x01\x00\n\x11\n\x08feature1\x12\x05\x1a\x03\n\x01\x00'])
).repeat().take(100)
tfrecord_writer = tf.data.experimental.TFRecordWriter('test.tfrecord')
tfrecord_writer.write(tf_examples_ds)

## load TF Transform and try to use it

tf_transform = tft.TFTransformOutput(transform_dir)
transformed_ds = tf.data.experimental.make_batched_features_dataset(
    file_pattern='test.tfrecord',
    batch_size=10,
    features=tf_transform.raw_feature_spec(),
).map(tf_transform.transform_raw_features)
list(transformed_ds.take(1))
```

Running this will throw `InvalidArgumentError: Unable to find FunctionDef for RaggedFromSparse_Assert_AssertGuard_true_352 in the registry. [Op:OptimizeDataset]`.

**Comment by zoyahav** [comment]
Which version of TFT and TF were you running with?
I'm seeing different errors:

> RuntimeError: __iter__() is only supported inside of tf.function or when eager execution is enabled.

**Comment by rafaelbarreto87** [comment]
As I mentioned this is for TF 2.x in eager mode. I created a Colab that you can just run here: https://colab.research.google.com/drive/17OygsqI4S2DoI5ii1UNYAtE1X_2DNQOi.

**Comment by zoyahav** [comment]
Your colab works if I modify L64-L70 to be:
```
with tf.Graph().as_default():
  tf_transform = tft.TFTransformOutput(transform_dir)
  transformed_ds = tf.data.experimental.make_batched_features_dataset(
      file_pattern='test.tfrecord',
      batch_size=10,
      features=tf_transform.raw_feature_spec(),
  ).map(tf_transform.transform_raw_features)
  x = tf.compat.v1.data.make_one_shot_iterator(
          transformed_ds).get_next()
  with tf.compat.v1.Session() as sess:
    print(sess.run(x))
```

But this means that we're not transforming the data eagerly.

Otherwise, starting with TFT 0.24 tft_beam.Context() accepts a parameter called force_tf_compat_v1 (defaults to True currently), which means that Transform is running in a TF v1 compatible mode.
If I set this to False in your colab, your original colab completes successfuly, the change is in L45:
```
    with tft_beam.Context(temp_dir=tempfile.mkdtemp(), force_tf_compat_v1=False):
```

**Comment by varshaan** [comment]
Alternately, you can use the TransformFeaturesLayer [1] to transform data eagerly. The colab also works if L64-70 is as follows:

tf_transform = tft.TFTransformOutput(transform_dir)
tft_layer = tf_transform.transform_features_layer()
transformed_ds = tf.data.experimental.make_batched_features_dataset(
    file_pattern='test.tfrecord',
    batch_size=10,
    features=tf_transform.raw_feature_spec(),
).map(tft_layer)
list(transformed_ds.take(1))  

[1] https://www.tensorflow.org/tfx/transform/api_docs/python/tft/TransformFeaturesLayer

**Comment by rafaelbarreto87** [comment]
@zoyahav @varshaan Thanks! Indeed, using `tft_layer = tf_transform.transform_features_layer()` solves the problem. I hadn't had much luck with `force_tf_compat_v1=False` though. By setting it to false, my pipeline starts to fail with:

```
ValueError: An error occured while trying to apply the transformation: "Dimensions 1 and 2 are not compatible".
          Batch instances: pyarrow.RecordBatch
```

I'm pretty sure I can debug and figure it out, but since it works without `force_tf_compat_v1=False`, I'll just go with the `tf_transform.transform_features_layer` solution for now.

Do we know the source of this? While I understand the compatibility issue, this issue was surprising to me that some operations work while others don't. It would be nice to have some documentation about it.

Thanks for the help.

**Comment by varshaan** [comment]
I fixed the dimension incompatibility in a recent commit [1]. It should be fixed with the next TFT release. I am glad using the transform features layer unblocked you though. 

[1] https://github.com/tensorflow/transform/commit/879f2345dcd6096104ae66027feacb099e228e66

**Comment by varshaan** [comment]
Closing this issue based on the last comment from Rafael. Please re-open if this is not fixed.
